WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupEpilogueFragment.java

```diff
@@ -0,0 +1,675 @@
+            options.setHideBottomControls(true);
+
+            UCrop.of(uri, Uri.fromFile(new File(context.getCacheDir(), "cropped.jpg")))
+                    .withAspectRatio(1, 1)
+                    .withOptions(options)
+                    .start(getActivity(), this);
+        }
+    }
+
+    protected void startGravatarUpload(final String filePath) {
+        if (!TextUtils.isEmpty(filePath)) {
+            File file = new File(filePath);
+
+            if (file.exists()) {
+                startProgress();
+
+                GravatarApi.uploadGravatar(file, mAccountStore.getAccount().getEmail(), mAccountStore.getAccessToken(),
+                        new GravatarApi.GravatarUploadListener() {
+                            @Override
+                            public void onSuccess() {
+                                endProgress();
+                                mPhotoUrl = GravatarUtils.fixGravatarUrl(mAccount.getAccount().getAvatarUrl(),
+                                        getResources().getDimensionPixelSize(R.dimen.avatar_sz_login_epilogue));
+                                mHeaderAvatar.setImageUrl(mPhotoUrl, ImageType.AVATAR);
+                                mHeaderAvatarAdd.setVisibility(View.GONE);
+                                mIsAvatarAdded = true;
+                            }
+
+                            @Override
+                            public void onError() {
+                                endProgress();
+                                showErrorDialogAvatar();
+                                AppLog.e(T.NUX, "Uploading image to Gravatar failed");
+                            }
+                        });
+            } else {
+                ToastUtils.showToast(getActivity(), R.string.error_locating_image, ToastUtils.Duration.SHORT);
+            }
+        } else {
+            ToastUtils.showToast(getActivity(), R.string.error_locating_image, ToastUtils.Duration.SHORT);
+        }
+    }
+
+    protected void undoChanges() {
+        mDisplayName = getArguments().getString(ARG_DISPLAY_NAME);
+        mEditTextDisplayName.setText(mDisplayName);
+        mUsername = getArguments().getString(ARG_USERNAME);
+        mEditTextUsername.setText(mUsername);
+        mPassword = "";
+        mInputPassword.getEditText().setText("");
+        updateAccountOrContinue();
+    }
+
+    protected void updateAccountOrContinue() {
+        if (changedDisplayName()) {
+            startProgress();
+            PushAccountSettingsPayload payload = new PushAccountSettingsPayload();
+            payload.params = new HashMap<>();
+            payload.params.put("display_name", mDisplayName);
+
+            if (changedPassword()) {
+                payload.params.put("password", mPassword);
+            }
+
+            mDispatcher.dispatch(AccountActionBuilder.newPushSettingsAction(payload));
+        } else if (changedPassword()) {
```

Comment:
I notice that the check for password and username modification is via `else if` blocks but I wonder, can't those be changed _in parallel_ to the display name?

Reply:
The display name and password can be changed in the same payload, but the username is different.  Dispatching those two payloads simultaneously makes it difficult to know when to continue since it's not guaranteed when each call will return.  That's why I chose to cascade them.

I did notice a logic error in the `updateAccountOrContinue()` method.  I updated it in 44382b51f13879e4bb86d37cb2c7fba626f294f1.